### ISSUE TITLE
HPA needs to handle discovery, controller credentials, privilege escalation protection

### ISSUE THREAD

**Issue by deads2k** [kubernetes/kubernetes#16177]
When integrating the HPA controller, we've hit a couple issues.

1. The HPA controller takes an APIGroup, APIVersion, Kind (this should probably have been resource) and attempts to call a `scale` operation on it.  Right now, that's based on a client building a URL.  That client needs to be able to perform API discovery operations in order to locate where the particular (group, version, resource) tuple lives.  Right now it doesn't.

1.  This intersects with authorization because the HPA controller needs to have valid credentials for every (group, version, resource) tuple that it can manipulate.  Given that the end goal is to have some APIGroups hosted in different locations (consider kube and openshift not living in the same API server), we need to have a way to initialize the HPA controller with correct credentials.

1.  Because the HPA controller may escalate the user's authority to actually perform the scale operation, the HPA REST endpoint should perform some authorization of the HPA resource when it is created.  One option is to do this via the `SubjectAccessReview` mechanism described here: https://github.com/kubernetes/kubernetes/pull/12209.  The remote API is probably needed because you'd want to authorize based on the authorization engine of the API server hosting the (grouip, version, resource).  In the short term, even accepting an `Authorizer` would get you part way.

1. Our APIInstaller infrastructure forces all requests for a particular version of a REST endpoint to use an exactly matching version (POSTs to api/v1/namespace/foo/resource must send a v1 object). This breaks the HPAController client because it cannot generically include every codec for every (group, kind) that it might want to scale.  https://github.com/kubernetes/kubernetes/issues/16177#issuecomment-151831609 Spawned https://github.com/kubernetes/kubernetes/issues/16500

@DirectXMan12 @ncdc @smarterclayton @kubernetes/kube-iam @kubernetes/rh-cluster-infra

**Comment by smarterclayton** [comment]
@lavalamp - this is essentially federation of cluster API groups that
aren't provided in the Kube code base, part of the reason I've been pushing
discovery use cases so hard.  The client discovers groups - some of those
groups are disconnected from the master, the HPA should be able to
autoscale them, etc.  I'll try to make sure there are relevant comments on
the appropriate design docs for this.

**Comment by lavalamp** [comment]
@deads2k @smarterclayton There is now a discovery client. I can't promise
it's final, but is there more that you need for point #1?

(my canonical example has been namespace controller until now, so now I
have two canonical examples.)

**Comment by deads2k** [comment]
> @deads2k @smarterclayton There is now a discovery client. I can't promise
> it's final, but is there more that you need for point #1?

Yes.  The current implementation doesn't allow for composing API groups from different sources.  If I'm hosting my resources in another location and I want one of my resources to be auto-scaled, I can't do it because the HPA controller (more precisely the contained client) has no way to determine where to submit the request.

**Comment by lavalamp** [comment]
OK, yeah, we have no solution for that at the moment. I think we have to
fix it in apiserver first.

@nikhiljindal This is a case we need to handle when registering 3rd party
API sources.

**Comment by deads2k** [comment]
Found another wrinkle.  Our `APIInstaller` infrastructure forces all requests for a particular version of a REST endpoint to use an exactly matching version (POSTs to `api/v1/namespace/foo/resource` must send a `v1` object).  This breaks the HPAController client because it cannot generically include every codec for every (group, kind) that it might want to scale.

We can try to fix this by allowing heterogeneous (group, version, kinds) and then forcing all scale endpoints to accept `extensions, +, scale`.   An unrecognized `extensions, version`, could be rejected with a header listing the recognized (group,version) for re-encoding.

**Comment by smarterclayton** [comment]
Can you describe exactly what is getting posted?

**Comment by smarterclayton** [comment]
HPAController is posting to `<thirdpartygroup>/<thirdpartyapiversion>/namespace/foo/resource`?

**Comment by deads2k** [comment]
> HPAController is posting to <thirdpartygroup>/<thirdpartyapiversion>/namespace/foo/resource

I propose posting an object of (extensions, latest-version, scale).  If the remote `thirdpartygroup` does not have that version, it replies back indicating which versions it does support and the HPA controller/client re-encodes and sends an object of  (extensions, specified-version, scale).

**Comment by smarterclayton** [comment]
Ok, so to further articulate the problem here, and why this is critically important:

1. HPA is a central component that should be able to reach out to generic endpoints
2. HPA can only know about a limited number of versions and conversions internally (anything compiled with kube)
3. We want HPA to be able to scale anything that results in pods generically, by virtue of invoking a generic Scale sub resource on endpoints
4. HPA will therefore have to be able to submit things *it* knows about to endpoints it *does not* know about.
5. Therefore an API server has to be able to accept the things *HPA* knows about (Scale, Kube group, v1) and convert it internally to its own recognized format (Scale, OpenShift/3rd party group, v2).
6. Therefore we need to relax the APIInstaller limitation and allow clients to convert.

@bgrant0607 @lavalamp i think this walks back the previous decision we had to not accept multiple types of input - there is no way for HPA to scale things it doesn't know about unless it can send things to a remote server that it knows about.

**Comment by bgrant0607** [comment]
One of the main purposes of many proposed subresources is polymorphism: support the same narrow API to apply the same semantic operation to multiple REST resources. Given that, I agree that, for polymorphic subresources, the apiserver should accept groups/versions that differ from the primary resource. We may eventually need a way to bound the set of supported groups/versions accepted at a particular path.

I believe HPA worked around this in testing by shadowing ReplicationController paths in extensions/v1beta1. I agree that's not a long-term solution.

For standard REST resources/operations, I think it's still better to be strict in order to prevent accidents.

Either way, if apiVersion and kind are omitted, we should be able to set them by default from the path, assuming such a resource exists.

**Comment by deads2k** [comment]
I'll start taking a look this afternoon.

**Comment by bgrant0607** [comment]
Hmm. GET would be a problem. There would need to be a way to specify the version.

**Comment by deads2k** [comment]
> Hmm. GET would be a problem. There would need to be a way to specify the version.

Request header and the server can choose a default if its not specified?

**Comment by smarterclayton** [comment]
It's possible the discovery API should provide an equivalence statement
(this object in this version maps to this object in the Kube base version)
for subresources that must be polymorphic

**Comment by bgrant0607** [comment]
@deads2k We so far haven't used request header in that way, and using 2 different methods (header and path) to choose api version seems confusing.

**Comment by bgrant0607** [comment]
I don't know that we can get a proper fix into 1.1. By when do you need a solution?

**Comment by lavalamp** [comment]
IMO, the group/version in the path must refer to the thing that you're writing or reading, or it shouldn't be in the path.

**Comment by bgrant0607** [comment]
Could you register paths at /apis/extensions/v1beta1/namespaces/{ns}/.../scale?

**Comment by deads2k** [comment]
> Could you register paths at /apis/extensions/v1beta1/namespaces/{ns}/.../scale?

That would mean that the list of resources under a given group could end up having resources from other API groups.  For example ` /apis/extensions/v1beta1/namespaces/{ns}/originresource/{name}/scale` would have an openshift resource.  That's not how I would expect to describe "I'm a resource that is a `Scaler`".  I would expect that determination to be made based on a discovery doc that indicated a resource with a subresource `scale` that can PUT and GET an (extensions, v1beta1, scale).

**Comment by DirectXMan12** [comment]
I think part of the issue is that scale acts less like a concrete resource, and more like a Go interface -- it's a promise that an object can be accessed using a common method -- in the case of a Go interface, that's methods, but in this case it's an API "object" that's calculated on demand, and propagates its changes back to the underlying object.

This strikes me as different from most of the rest of the API.

**Comment by lavalamp** [comment]
HPA can just GET `/apis/extensions/v1beta1/namespaces/{ns}/originresource/{name}/scale` to see if it's supported or not. Longer term we should just include a URL, e.g.:

```
...
scale = "/apis/extensions/v1beta1/namespaces/{ns}/originresource/{name}/scale"
...
```

I'm not sure how you're going to get HPA to know it should point at OpenShift instead of k8s apiserver.

**Comment by davidopp** [comment]
@kubernetes/autoscaling

**Comment by davidopp** [comment]
Assigning to @lavalamp and marking 1.1

**Comment by bgrant0607** [comment]
Can we please split this into separate github issues for each distinct technical issue?

**Comment by davidopp** [comment]
> Can we please split this into separate github issues for each distinct technical issue?

+1

at the very least, need to separate "hits anyone using out-of-the-box k8s" vs. "hits anyone using systems built on top of k8s that add resources"

@lavalamp said the first category is empty -- I had put this in 1.1 because I thought it was non-empty, but apparently I misunderstood.

**Comment by DirectXMan12** [comment]
> @lavalamp said the first category is empty

One could argue that having all kube objects with a scale subresource have a corresponding url in 'extensions/v1beta1' is a bit strange, since the rest of the object is not accessible at that URL.  It's doesn't *prevent* core Kube from working, but perhaps it's something to consider for the future.

**Comment by lavalamp** [comment]
I think the specific CSI issues here are bigger than we can reasonably get fixed for 1.1.

**Comment by smarterclayton** [comment]
That's not a composition pattern - that's an inheritance pattern, which
means it's inherently limiting clients.

Hpa is a client just like kubectl, it reads discovery doc, finds a scale
subresource, and targets it.

On Oct 28, 2015, at 2:47 PM, Daniel Smith <notifications@github.com> wrote:

HPA can just GET
/apis/extensions/v1beta1/namespaces/{ns}/originresource/{name}/scale to see
if it's supported or not. Longer term we should just include a URL, e.g.:

...
scale = "/apis/extensions/v1beta1/namespaces/{ns}/originresource/{name}/scale"
...

I'm not sure how you're going to get HPA to know it should point at
OpenShift instead of k8s apiserver.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/16177#issuecomment-151949577>
.

**Comment by smarterclayton** [comment]
I agree, but I also think we are not succeeding at some of our core goals
for Kube in this space w.r.t building a system that is suitable for
extension.  We need to spend a bit more time coming to design agreement on
some of these patterns in 1.2 - I'd like to ensure we can learn from the
pain points in 1.1 and apply those lessons for future work.  Having
everyone on the same page about some of these issues is really important -
we're getting there, but it's not quite a reality yet.

On Oct 28, 2015, at 5:05 PM, Daniel Smith <notifications@github.com> wrote:

I think the specific CSI issues here are bigger than we can reasonably get
fixed for 1.1.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/16177#issuecomment-151989691>
.

**Comment by deads2k** [comment]
Spawned https://github.com/kubernetes/kubernetes/issues/16500 for item 4, which we all seem to agree on

**Comment by ncdc** [comment]
cc @piosz

**Comment by jszczepkowski** [comment]
CC @jszczepkowski

**Comment by aronchick** [comment]
@deads2k @smarterclayton You guys have this on your blocker list, can either of you take this across the finish for 1.2?

**Comment by ncdc** [comment]
@aronchick will do. @deads2k, do you have a summary of what remaining outstanding work there is related to this issue?

**Comment by deads2k** [comment]
> @aronchick will do. @deads2k, do you have a summary of what remaining outstanding work there is related to this issue?

Item 1 got picked up, sort of (you specify a group in the `apiVersion` field to specify a group) and I haven't looked recently to see if the Kind versus Resource distinction is possible.

Everything else is still outstanding.

**Comment by ncdc** [comment]
@deads2k are items 1 and 2 must-haves for

**Comment by davidopp** [comment]
@ncdc @deads2k Feature complete deadline for 1.2 is Friday and it doesn't seem like this is going to be finished by then. OK to bounce it from the 1.2 milestone to the next release?

**Comment by ncdc** [comment]
@deads2k WDYT?

**Comment by deads2k** [comment]
> @deads2k WDYT?

I think what we've agreed to in https://github.com/kubernetes/kubernetes/issues/18528 is sufficient for 1.2.

**Comment by mwielgus** [comment]
This is probably covered by RBAC now. 
@DirectXMan12, can we close this issue?

**Comment by DirectXMan12** [comment]
I think this may be mostly solved.  Not all of the issues that @deads2k mentioned are quite fixed though (giving write access to HPA is effectively equivalent to giving access to */scale, for instance, and we still don't do discovery quite properly), but I think we can open separate issues if we need to.

**Comment by ncdc** [comment]
@DirectXMan12 let's close this and create separate individual issues for any remaining work items?

**Comment by DirectXMan12** [comment]
sure
